datafusion/optimizer/src/type_coercion.rs

```diff
@@ -116,18 +117,61 @@ impl ExprRewriter for TypeCoercionRewriter {
                     }
                 }
             }
+            Expr::ScalarUDF { fun, args } => {
+                let new_expr = coerce_arguments_for_signature(
+                    args.as_slice(),
+                    &self.schema,
+                    &fun.signature,
+                )?;
+                Ok(Expr::ScalarUDF {
+                    fun: fun.clone(),
+                    args: new_expr,
+                })
+            }
             _ => Ok(expr),
         }
     }
 }
 
+/// Returns `expressions` coerced to types compatible with
+/// `signature`, if possible.
+///
+/// See the module level documentation for more detail on coercion.
+pub fn coerce_arguments_for_signature(
```

Comment:
I feel like this code should already exist somewhere and it would be great to consolidate into a single implementation rather than have multiple implementations around

Reply:
Yes, this is adapted from the `coerce` method in the `physical-expr` crate that operates on physical expressions rather than logical expressions.

Reply:
Do we need to public this method?
I think it's better to make it private.